### PR TITLE
json-rpc: translate NULL into "null" instead of "(null)".

### PR DESCRIPTION
**Pull request by bitonic-cjp**
It turns out that #957 is insufficient to solve #908 . This fix is needed to really say "id": null instead of "id": (null).

This is probably an incomplete fix: it would be better to make sure that **all** cases of NULL are translated into "null". However, I didn't want to touch the tal_fmt implementation. What would be a good approach? make a json_fmt wrapper around tal_fmt to do this? I didn't do this yet: since I'm not too familiar with the code base yet, I didn't want to make too large changes without any guidance.

### CODE REVIEW

**Comment by ZmnSCPxj** [comment]
`tal_fmt` is a wrapper around `vsnprintf`, and if you want to be pedantic, a `"%s"` format string should not be given a `NULL` as that will lead to undefined behavior.  It is simply that the behavior in glibc is to print `"(null)"` in that case; we cannot rely on all systems doing that.

**Comment by ZmnSCPxj** [comment]
I believe this is the point which we should modify:

https://github.com/ElementsProject/lightning/blob/73cda2f2ae027f3c2c06b2fe1b15f18f28487a45/lightningd/jsonrpc.c#L687-L689

It should be `json_command_malformed(jcon, "null", "Invalid token in json input");`

Then inside both `connection_complete_error` and `connection_complete_ok` we should have `assert(id != NULL)` to catch this in the future.

**Comment by bitonic-cjp** [comment]
Agreed: that seems to be somewhat less ugly than what I made.

Could we somehow ensure that NULL is never passed to tal_fmt through another route? I find it scary to have stumbled upon something that is actually undefined behavior.

**Comment by ZmnSCPxj** [comment]
Wait for @practicalswift to do [regular](https://github.com/ElementsProject/lightning/pull/1032) [undefined](https://github.com/ElementsProject/lightning/pull/953) [behavior](https://github.com/ElementsProject/lightning/pull/640) [checking](https://github.com/ElementsProject/lightning/pull/537) has always been my personal technique, hahaha.

**Comment by rustyrussell** [comment]
Was about to suggest  the same as @ZmnSCPxj : let's kill the root cause, and assert() that it never happens again.

**Comment by ZmnSCPxj** [comment]
On a more serious note, @bitonic-cjp: undefined behavior is mostly caught by code review and sometimes by automated tools that scan for them, and `valgrind` can catch some bits of undefined behavior related to usage of memory.  @practicalswift might have a tool or two that scans for undefined behavior (and has requested pulls to add such linting to the python parts of our codebase, for instance).  I have caught a few bugs myself (e.g. subdaemons blocking `SIGCHLD` when they should be blocking `SIGPIPE`, `waitanyinvoice` blocking even with an invoice to report or reporting an invoice twice) simply from code review.  So mostly.... with enough eyes, all bugs are shallow.  Hopefully with more people supporting development we can get more bugs found and fixed.

**Comment by bitonic-cjp** [comment]
I did what ZmnSCPxj said; now I'm checking other tal_fmt calls.

**Comment by ZmnSCPxj** [comment]
ACK 59b7f9e407ecb0511ed1c46708593b97fc984a04

File: lightningd/jsonrpc.c [hunk 1]
```diff
@@ -263,6 +263,8 @@ static void connection_complete_ok(struct json_connection *jcon,
 				   const char *id,
 				   const struct json_result *result)
 {
+    assert(id != NULL);
```
Comment (practicalswift) [inline lightningd/jsonrpc.c]: Nit: Make sure the indentation matches the surrounding code. Same below.

Comment (bitonic-cjp) [reply lightningd/jsonrpc.c]: Agreed; I already discovered this myself, and was going to fix it.

**Comment by cdecker** [comment]
ACK 500db24

**Pull request merged**